language-hask-source/src/Language/Hask/Parsed/AST.hs

```diff
@@ -0,0 +1,575 @@
+  , -- * Types
+    Type (..)
+  , -- * Parsed names
+    ModuleName (..)
+  , Qualified (..)
+  , VariableName (..)
+  , ConstructorName (..)
+  , Name (..)
+    -- * Notes
+    -- $ParsedTypeImportLists
+    -- $TypesInDeclarationHeads
+    -- $BindingsInInstances
+  ) where
+
+import Data.Kind qualified as Kind
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Text (Text)
+import Numeric.Natural
+
+-------------------------------------------------------------------------------
+-- | Parsed Haskell module.
+--
+-- > module $moduleId ($moduleExports) where
+-- >
+-- > $moduleImports
+-- >
+-- > $moduleDeclarations
+data Module = Module{
+    moduleName         :: ModuleName
+  , moduleExports      :: [Export]
+  , moduleImports      :: [ImportDeclaration]
+  , moduleDeclarations :: [Declaration]
+  } deriving stock Show
+
+-- | Parsed export from current module.
+data Export
+  = VariableExport (Qualified VariableName)
+    -- ^ Export of a simple variable.
+  | TypeExport (Qualified ConstructorName) (Maybe TypeImportList)
+    -- ^ Export of a type.
+    --
+    -- > $1
+    --
+    -- or
+    --
+    -- > $1 ($2)
+    --
+    -- See [Parsed type import/export lists](#ParsedTypeImportLists).
+  | ModuleExport ModuleName
+    -- ^ Export of a module.
+    --
+    -- > module $1
+  deriving stock Show
+
+-- | Parsed import declaration.
+--
+-- > import $importQualified $importName $importAlias $importHiding ($importList)
+data ImportDeclaration = ImportDeclaration{
+    importQualified :: Bool             -- ^ Presence of @qualified@ keyword.
+  , importName      :: ModuleName
+  , importAlias     :: Maybe ModuleName -- ^ Presense of @as $_@ syntax.
+  , importHiding    :: Bool             -- ^ Presence of @hiding@ keyword.
+  , importList      :: [Import]
+  } deriving stock Show
+
+-- | Parsed import.
+data Import
+  = VariableImport VariableName
+    -- ^ Import of a simple variable.
+  | TypeImport ConstructorName (Maybe TypeImportList)
+    -- ^ Import of a type.
+    --
+    -- > $1
+    --
+    -- or
+    --
+    -- > $1 ($2)
+    --
+    -- See [Parsed type import/export lists](#ParsedTypeImportLists).
+  deriving stock Show
+
+-- | Parsed import/export list of a type. You may be interested in
+-- [Parsed type import/export lists](#ParsedTypeImportLists).
+data TypeImportList
+  = TypeImportAll
+    -- ^ Import/export of all members in scope.
+    --
+    -- > (..)
+  | TypeImportList [Either VariableName ConstructorName]
+    -- ^ Import/export of explicitly named members.
+    --
+    -- > ($1)
+  deriving stock Show
+
+-- $ParsedTypeImportLists
+-- == Parsed type import/export lists #ParsedTypeImportLists#
+-- 'TypeImportList's contain both class and datatype imports/exports, because
+-- we can't yet distinguish those properly - that's up to the renamer.
+
+-------------------------------------------------------------------------------
+-- | Parsed top-level declaration.
+data Declaration
+  = TypeSynonym Type Type
```

Comment:
Could we do something more well typed here for example I would expect something like:

```haskell
TypeSynonym Name [Vars] Type -- corresponding to type Name v1 ... vn = type
```

Reply:
I guess it depends on whether we want to report
```
unexpected <random thing>, expected type variable
```
or
```
type synonym <type synonym> is malformed
<random thing> appears as an argument, but arguments to type synonyms must be type variables
```
later when converting to more concrete type.